### PR TITLE
Etl fhir

### PR DESCRIPTION
**Pull request by kairstenfay**
Ready for review! I have a couple of lingering TODOs related to using test data plus some leftover conceptual questions. 99% of the work here is done!

### CODE REVIEW

**Comment by kairstenfay** [comment]
Ready for review!

**Comment by kairstenfay** [comment]
I just removed an unused import from `fhir.py`

File: lib/id3c/cli/command/etl/redcap_det.py [hunk 1]
```diff
@@ -116,11 +116,11 @@ def decorated(*args, db: DatabaseSession, log_output: bool, **kwargs):
...
-                    if incomplete_instruments:
-                        LOG.debug(f"The following required instruments «{incomplete_instruments}» are not yet marked complete. " + \
-                                  f"Skipping REDCap DET {det.id}")
-                        mark_skipped(db, det.id, etl_id)
-                        continue
+                    # if incomplete_instruments:
+                    #     LOG.debug(f"The following required instruments «{incomplete_instruments}» are not yet marked complete. " + \
+                    #               f"Skipping REDCap DET {det.id}")
+                    #     mark_skipped(db, det.id, etl_id)
+                    #     continue
```
Comment (joverlee521) [inline lib/id3c/cli/command/etl/redcap_det.py]: I think this snuck into your commit while you were testing?

File: lib/id3c/cli/command/etl/fhir.py [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+    """
+    for result in report.result:
+        observation = result.resolved(Observation)
+
+        # Most of the time we expect to see existing targets so a
+        # select-first approach makes the most sense to avoid useless
+        # updates.
+        target = find_or_create_target(db,
+            identifier  = matching_system_code(observation.code, TARGET_SYSTEM),
+            control     = False)  # TODO what do we expect here? Do we expect more controls?
```
Comment (joverlee521) [inline lib/id3c/cli/command/etl/fhir.py]: I think default to `False` is fine? We can manually update controls if we get new ones?

File: lib/id3c/cli/command/etl/fhir.py [hunk 3]
```diff
@@ -0,0 +1,639 @@
...
+
+        upsert_presence_absence(db,
+            identifier = f'{barcode}/{observation.device.identifier.value}',  # TODO is this correct use of assay?
```
Comment (joverlee521) [inline lib/id3c/cli/command/etl/fhir.py]: This identifier won't work if you have multiple results for the same barcode on the same device. Maybe include the target identifier?

File: lib/id3c/cli/command/etl/fhir.py [hunk 4]
```diff
@@ -0,0 +1,639 @@
...
+            sample_id = sample_id,
+            target_id = target.id,
+            present = True,
```
Comment (joverlee521) [inline lib/id3c/cli/command/etl/fhir.py]: This should come from `observation.valueBoolean`

File: lib/id3c/cli/command/etl/__init__.py [hunk 5]
```diff
@@ -419,6 +420,128 @@ def upsert_encounter_location(db: DatabaseSession,
...
+    """
+    Upsert presence_absence by its *identifier*.
+
+    Confirmed with Samplify that their numeric identifier for each test is stable
+    and persistent.
+    """
+    LOG.debug(f"Upserting presence_absence «{identifier}»")
+
+    data = {
+        "identifier": f"NWGC_{identifier}",
```
Comment (joverlee521) [inline lib/id3c/cli/command/etl/__init__.py]: Just wanted to flag that in the most recent changes to `etl presence-absence`, this got updated to 
```
"identifier": identifier,
```

File: lib/id3c/cli/command/etl/fhir.py (hunk 4 above)
Comment (kairstenfay) [reply lib/id3c/cli/command/etl/fhir.py]: oh yes, good catch thank you

File: lib/id3c/cli/command/etl/redcap_det.py (hunk 1 above)
Comment (kairstenfay) [reply lib/id3c/cli/command/etl/redcap_det.py]: yeah, looks like it. thanks

File: lib/id3c/cli/command/etl/fhir.py [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+        identifier  = identifier(patient, f"{INTERNAL_SYSTEM}/individual"),
+        sex         = sex(patient))
+
+
+def process_encounter_site(db: DatabaseSession, encounter: Encounter) -> Optional[Any]:
+    """
+    Returns a found or created ``site`` (per ID3C encounter-relation
+    definitions) using data from the given *encounter*'s linked Location
+    Resources.
+    """
```
Comment (kairstenfay) [inline lib/id3c/cli/command/etl/fhir.py]: I should add a not in the docstring that it returns the first site that's found

**Comment by kairstenfay** [comment]
[Core ID3C ETL routine to import FHIR bundles](https://trello.com/c/MGHNJu4V/246-core-id3c-etl-routine-to-import-fhir-bundles)

File: lib/id3c/cli/command/etl/fhir.py [hunk 7]
```diff
@@ -0,0 +1,639 @@
...
+    unique_codes = list(set(codes))
+    if len(unique_codes) > 1:
+        raise Exception(f"Expected only one Location code. Instead received {unique_codes}.")
+
+    return unique_codes[0]
+
+
+def process_sample(db: DatabaseSession, barcode: str) -> Any:
+    """ Given a *barcode*, returns its matching sample from ID3C. """
+    sample = find_sample(db, barcode)
```
[the remark this answers is not in the document]
Comment (tsibley) [reply lib/id3c/cli/command/etl/fhir.py]: FWIW, in other ETLs, we find samples by either their `identifier` or `collection_identifier`, without caring which is which.

File: lib/id3c/cli/command/etl/fhir.py (hunk 2 above)
Comment (tsibley) [reply lib/id3c/cli/command/etl/fhir.py]: That sounds ok to me.

**Comment by tsibley** [comment]
(Rebased this onto the latest `redcap-etl` and repushed)

**Comment by kairstenfay** [comment]
This PR was closed automatically by GitHub. Please see https://github.com/seattleflu/id3c/pull/90 for a continued conversation.

**Pull request closed without merging**